### PR TITLE
Add pthread_setaffinity_np method to support musl builds

### PR DESCRIPTION
**Pull request by r-burns**
Some platforms, such as musl, do not have support for
the nonportable pthread_attr_setaffinity_np function.
Some of those without may be able to fall back to
pthread_setaffinity_np.

We can directly check for support using CMake's CheckSymbolExists,
and enable/disable this functionality accordingly.

Fixes https://github.com/awslabs/aws-c-common/issues/754

(By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.)

### CODE REVIEW

File: source/posix/thread.c
Comment (JonathanHenson) [inline source/posix/thread.c:163]: This code is actually vital if you want to hit full NIC entitlement on numa architectures. Are there any ways we can check for the alternative variants and use those where we can? I suspect a lot of Rust and C++ users link against musl and wouldn't like to see a massive performance reduction because of it.

Comment (r-burns) [reply source/posix/thread.c:163]: Do you mean like falling back to `pthread_setaffinity_np`? I think that extension is supported on musl.

Comment (JonathanHenson) [reply source/posix/thread.c:163]: Yes, that would work.

Comment (JonathanHenson) [reply source/posix/thread.c:163]: In fact, that might also work for BSD variants.

Comment (r-burns) [reply source/posix/thread.c]: Okay, the patch is a bit more complicated with this fallback mechanism, but it's definitely valuable to add support for musl and maybe even BSD. I don't have a BSD setup to test on but maybe I can try it out in a VM later this week. Would appreciate if you could take a look at the new revision I pushed when you have the time.

**Review by JonathanHenson — APPROVED** [review]

File: source/posix/thread.c
Comment (JonathanHenson) [reply source/posix/thread.c]: this looks a good deal better. Thanks! now we're going to have to work around github actions and move this to a branch so CI will run. I'll link it here.

File: cmake/AwsThreadAffinity.cmake
Comment (JonathanHenson) [inline cmake/AwsThreadAffinity.cmake:25]: can we make these checks mutually exclusive so we don't accidentally use both functions if for some reason an OS distribution does something weird and wonky?

File: source/posix/thread.c
Comment (JonathanHenson) [inline source/posix/thread.c:213]: can we make these mutually exclusive, just in case?

Comment (JonathanHenson) [reply source/posix/thread.c:213]: I would say the cmake check being hardened would be enough, but over time we're going to have a few builds that don't go through cmake and instead go through swift.package files and build.rs systems. Those will be emulating everything the cmake currently does and I'd like to prevent problems here if we get that part wrong.

Comment (r-burns) [reply source/posix/thread.c:213]: Thanks, I was wondering about this while adding these macros. I can see two possible approaches:

1. Add checks such as
#if (defined(X) && defined(Y)
#error
Probably simplest but would combinatorially increase if more mutually exclusive methods are added
2. Treat the method macros like an enum to be matched against possible values, e.g.
#if AWS_AFFINITY_METHOD == AWS_AFFINITY_METHOD_PTHREAD_ATTR

Comment (JonathanHenson) [reply source/posix/thread.c:213]: I _think_ N is very small here, so I'm not particularly worried about 1. 2. is fine too. I leave it up to you, whatever seems cleanest and easiest to grok for maintainers, and still prevents a misconfiguration step from borking the builds.

File: source/posix/thread.c
Comment (r-burns) [inline source/posix/thread.c:29]: This intentionally fails if the method is not specified. It could fall back to NONE, but I considered this more robust as non-cmake build systems can't accidentally build with degraded performance if they are not aware of this.

File: source/posix/thread.c
Comment (r-burns) [inline source/posix/thread.c:41]: Possibly redundant, but extra checking. Can remove if deemed unnecessary.

File: source/posix/thread.c
Comment (r-burns) [reply source/posix/thread.c]: I ended up going with method 2, which is a bit more verbose but seems more self-documenting and maintainable.

File: source/posix/thread.c
Comment (ColdenCullen) [reply source/posix/thread.c:41]: It's not redundant, because in the above check you only check for existence, not value.

File: source/posix/thread.c
Comment (ColdenCullen) [inline source/posix/thread.c:239]: This is the same as above, you could move this out of the #if blocks.

Comment (r-burns) [reply source/posix/thread.c:239]: It needs to be guarded somehow though, I think because not all platforms have support for CPU_ZERO/SET. This was previously handled by the
```cpp
#if !defined(__MACH__) && !defined(__ANDROID__)
```

Comment (ColdenCullen) [reply source/posix/thread.c:239]: Ah good point.

**Comment by r-burns** [comment]
Friendly ping :)

**Review by JonathanHenson — APPROVED** [review]

**Comment by JonathanHenson** [comment]
closing this one so we can run CI and merge. The new PR is here:
https://github.com/awslabs/aws-c-common/pull/766

**Comment by JonathanHenson** [comment]
See the PR runs on this. We missed apple OS family which doesn't have any mechanism to set thread affinity since it's always in interleave mode anyways.

**Comment by r-burns** [comment]
Should I just push here so you can merge into the new PR branch?

**Comment by JonathanHenson** [comment]
Yeah.... unfortunately I'll have to manually merge it due to github security rules and the way github actions works at the moment.

**Comment by r-burns** [comment]
Oh right, the commits don't show up here now that the PR is closed. Anyway, I pushed the fixes to this branch.

**Comment by JonathanHenson** [comment]
okay, your changes have been merged. Thanks!

**Comment by r-burns** [comment]
Awesome, thanks! Will look forward to it in the next release :)

**Pull request closed without merging**
